cli/flox-rust-sdk/src/providers/publish.rs

```diff
@@ -94,55 +98,77 @@ pub struct CheckedBuildMetadata {
+        client: &Client,
+        floxhub_token: &FloxhubToken,
     ) -> Result<(), PublishError> {
         // Get metadata from the environment, like locked URLs.
-        let (_env_meta, _build_meta) = match environment {
-            PublishEnvironment::Managed(_env) => return Err(PublishError::UnsupportedEnvironment),
-            PublishEnvironment::Path(env) => (
-                check_environment_metadata(flox, &env)?,
-                check_build_metadata(&env, package)?,
-            ),
-        };
 
-        let _package_name = package;
-        let _catalog_name = match &flox.floxhub_token {
-            Some(token) => token.handle().to_string(),
-            None => return Err(PublishError::Unauthenticated),
+        let catalog_name = floxhub_token.handle().to_string();
+
+        // The create package service call will create the user's own catalog
+        // if not already created, and then create (or return) the package noted
+        // returning either a 200 or 201.  Either is ok here, as long as it's not an error.
+        client
+            .create_package(&catalog_name, &self.build_meta.package)
+            .await
+            .map_err(|e| PublishError::CatalogError(Box::new(e)))?;
+
+        let outputs = Outputs(
+            self.build_meta
+                .outputs
+                .clone()
+                .into_iter()
+                .map(|o| Output {
+                    name: o.name,
+                    store_path: o.store_path,
+                })
+                .collect(),
+        );
+
+        let build_info = UserBuildInfo {
+            derivation: UserDerivationInfo {
+                broken: Some(false),
+                description: "".to_string(),
+                drv_path: self.build_meta.drv_path.clone(),
+                license: None,
+                name: self.build_meta.package.to_string().to_owned(),
+                outputs,
+                outputs_to_install: None,
+                pname: Some(self.build_meta.package.to_string()),
+                system: self.build_meta.system,
+                unfree: None,
+                version: None,
+            },
+            locked_base_catalog_url: self.env_meta.build_repo_ref.url.clone(),
+            locked_url: self.env_meta.base_catalog_ref.as_ref().unwrap().url.clone(),
         };
-        let _version: Option<String> = None;
-        let _description: Option<String> = None;
-
-        // Uses client to...
-        // ... check access to the catalog
-        // ... check presence of and create the package if needed
-        // ... publish the build info
-        // publish_to_catalog()
+        client
+            .publish_build(&catalog_name, &self.build_meta.package, &build_info)
+            .await
+            .map_err(|e| PublishError::CatalogError(Box::new(e)))?;
+
         Ok(())
     }
 }
 
 /// Collect metadata needed for publishing that is obtained from the build output
-fn check_build_metadata(
+pub fn check_build_metadata(
```

Comment:
did these need to become public?

EDIT: I see they are being created first and then pushed into publish() 🤔

Reply:
The `provider` struct didn't have anything in it, and combining that with limiting where `flox` was used as a whole, I thought the checked data could be done up front and used to create the provider, and then the publish call would have what it needed without needing `flox` at that stage.   Happy to discuss when you have time, but I'll push this in, this time. I don't feel strongly on the structure.  Consistency is good.  I'll make a note for this, along with the async comment to revisit.